# TV Script Generation

In this project, you'll generate your own [Seinfeld](https://en.wikipedia.org/wiki/Seinfeld) TV scripts using RNNs.  You'll be using part of the [Seinfeld dataset](https://www.kaggle.com/thec03u5/seinfeld-chronicles#scripts.csv) of scripts from 9 seasons.  The Neural Network you'll build will generate a new ,"fake" TV script, based on patterns it recognizes in this training data.

## Get the Data

The data is already provided for you in `./data/Seinfeld_Scripts.txt` and you're encouraged to open that file and look at the text. 
>* As a first step, we'll load in this data and look at some samples. 
* Then, you'll be tasked with defining and training an RNN to generate a new script!

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# load in data
import helper
data_dir = './data/Seinfeld_Scripts.txt'
text = helper.load_data(data_dir)

## Explore the Data
Play around with `view_line_range` to view different parts of the data. This will give you a sense of the data you'll be working with. You can see, for example, that it is all lowercase text, and each new line of dialogue is separated by a newline character `\n`.

In [2]:
view_line_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in text.split()})))

lines = text.split('\n')
print('Number of lines: {}'.format(len(lines)))
word_count_line = [len(line.split()) for line in lines]
print('Average number of words in each line: {}'.format(np.average(word_count_line)))

print()
print('The lines {} to {}:'.format(*view_line_range))
print('\n'.join(text.split('\n')[view_line_range[0]:view_line_range[1]]))

Dataset Stats
Roughly the number of unique words: 46367
Number of lines: 109233
Average number of words in each line: 5.544240293684143

The lines 0 to 10:
jerry: do you know what this is all about? do you know, why were here? to be out, this is out...and out is one of the single most enjoyable experiences of life. people...did you ever hear people talking about we should go out? this is what theyre talking about...this whole thing, were all out now, no one is home. not one person here is home, were all out! there are people trying to find us, they dont know where we are. (on an imaginary phone) did you ring?, i cant find him. where did he go? he didnt tell me where he was going. he must have gone out. you wanna go out you get ready, you pick out the clothes, right? you take the shower, you get all ready, get the cash, get your friends, the car, the spot, the reservation...then youre standing around, what do you do? you go we gotta be getting back. once youre out, you wanna get back! y

---
## Implement Pre-processing Functions
The first thing to do to any dataset is pre-processing.  Implement the following pre-processing functions below:
- Lookup Table
- Tokenize Punctuation

### Lookup Table
To create a word embedding, you first need to transform the words to ids.  In this function, create two dictionaries:
- Dictionary to go from the words to an id, we'll call `vocab_to_int`
- Dictionary to go from the id to word, we'll call `int_to_vocab`

Return these dictionaries in the following **tuple** `(vocab_to_int, int_to_vocab)`

In [3]:
import problem_unittests as tests

from collections import Counter
import numpy as np

def create_lookup_tables(text):
    """
    Create lookup tables for vocabulary
    :param text: The text of tv scripts split into words
    :return: A tuple of dicts (vocab_to_int, int_to_vocab)
    """
    # TODO: Implement Function
    unique_words = Counter(text).most_common()
    unique_word_indexes = {pair[0]: indexer for indexer, pair in enumerate(unique_words)}
    vocab_to_int = {word: unique_word_indexes[word] for word in text}
    int_to_vocab = {unique_word_indexes[word] : word for word in text}
    
    print(vocab_to_int)
    print(int_to_vocab)
    
    # return tuple
    return (vocab_to_int, int_to_vocab)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_create_lookup_tables(create_lookup_tables)

{'moe_szyslak': 0, "moe's": 20, 'tavern': 21, 'where': 22, 'the': 7, 'elite': 23, 'meet': 24, 'to': 5, 'drink': 6, 'bart_simpson': 25, 'eh': 26, 'yeah': 8, 'hello': 27, 'is': 28, 'mike': 1, 'there': 29, 'last': 30, 'name': 9, 'rotch': 2, 'hold': 31, 'on': 10, "i'll": 32, 'check': 33, 'hey': 11, 'has': 34, 'anybody': 35, 'seen': 36, 'lately': 37, 'listen': 38, 'you': 3, 'little': 39, 'puke': 40, 'one': 12, 'of': 41, 'these': 42, 'days': 43, "i'm": 13, 'gonna': 14, 'catch': 44, 'and': 45, 'carve': 46, 'my': 15, 'your': 4, 'back': 47, 'with': 48, 'an': 49, 'ice': 50, 'pick': 51, 'whats': 52, 'matter': 53, 'homer': 16, "you're": 54, 'not': 17, 'normal': 55, 'effervescent': 56, 'self': 57, 'homer_simpson': 58, 'i': 59, 'got': 60, 'problems': 18, 'moe': 61, 'give': 62, 'me': 63, 'another': 64, 'should': 19, 'forget': 65, 'barney_gumble': 66, 'only': 67, 'enhance': 68, 'social': 69, 'skills': 70}
{0: 'moe_szyslak', 20: "moe's", 21: 'tavern', 22: 'where', 7: 'the', 23: 'elite', 24: 'meet', 5: 

### Tokenize Punctuation
We'll be splitting the script into a word array using spaces as delimiters.  However, punctuations like periods and exclamation marks can create multiple ids for the same word. For example, "bye" and "bye!" would generate two different word ids.

Implement the function `token_lookup` to return a dict that will be used to tokenize symbols like "!" into "||Exclamation_Mark||".  Create a dictionary for the following symbols where the symbol is the key and value is the token:
- Period ( **.** )
- Comma ( **,** )
- Quotation Mark ( **"** )
- Semicolon ( **;** )
- Exclamation mark ( **!** )
- Question mark ( **?** )
- Left Parentheses ( **(** )
- Right Parentheses ( **)** )
- Dash ( **-** )
- Return ( **\n** )

This dictionary will be used to tokenize the symbols and add the delimiter (space) around it.  This separates each symbols as its own word, making it easier for the neural network to predict the next word. Make sure you don't use a value that could be confused as a word; for example, instead of using the value "dash", try using something like "||dash||".

In [4]:
def token_lookup():
    """
    Generate a dict to turn punctuation into a token.
    :return: Tokenized dictionary where the key is the punctuation and the value is the token
    """
    # TODO: Implement Function
    punctuation_to_unique_name = {
        ".": "||period||",
        ",": "||comma||",
        "\"": "||quotation_mark||",
        ";": "||semicolon||",
        "!": "||exclamation_mark||",
        "?": "||question_mark||",
        "(": "||left_parentheses||",
        ")": "||right_parentheses||",
        "-": "||dash||",
        "\n": "||return||"
    }
        
    return punctuation_to_unique_name

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_tokenize(token_lookup)

Tests Passed


## Pre-process all the data and save it

Running the code cell below will pre-process all the data and save it to file. You're encouraged to lok at the code for `preprocess_and_save_data` in the `helpers.py` file to see what it's doing in detail, but you do not need to change this code.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# pre-process training data
helper.preprocess_and_save_data(data_dir, token_lookup, create_lookup_tables)

{'this': 24, 'is': 22, 'out': 47, '||period||': 1, 'and': 17, 'one': 82, 'of': 20, 'the': 6, 'single': 1252, 'most': 545, 'enjoyable': 8782, 'experiences': 7189, 'life': 241, 'people': 149, 'did': 84, 'you': 4, 'ever': 200, 'hear': 238, 'talking': 208, 'about': 58, 'we': 55, 'should': 135, 'go': 64, '||question_mark||': 3, 'what': 18, 'theyre': 677, 'whole': 220, 'thing': 126, '||comma||': 2, 'were': 121, 'all': 50, 'now': 86, 'no': 26, 'home': 289, 'not': 45, 'person': 374, 'here': 62, '||exclamation_mark||': 10, 'there': 76, 'are': 48, 'trying': 271, 'to': 8, 'find': 248, 'us': 191, 'they': 65, 'dont': 204, 'know': 27, 'where': 128, '||left_parentheses||': 11, 'on': 28, 'an': 98, 'imaginary': 2744, 'phone': 172, '||right_parentheses||': 12, 'ring': 1366, 'i': 5, 'cant': 388, 'him': 69, 'he': 35, 'didnt': 526, 'tell': 97, 'me': 25, 'was': 41, 'going': 80, 'must': 337, 'have': 39, 'gone': 572, 'wanna': 216, 'get': 51, 'ready': 427, 'pick': 323, 'clothes': 558, 'right': 57, 'take': 103,

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

int_text, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()

## Build the Neural Network
In this section, you'll build the components necessary to build an RNN by implementing the RNN Module and forward and backpropagation functions.

### Check Access to GPU

In [2]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch

# Check for a GPU
train_on_gpu = torch.cuda.is_available()
if not train_on_gpu:
    print('No GPU found. Please use a GPU to train your neural network.')

## Input
Let's start with the preprocessed input data. We'll use [TensorDataset](http://pytorch.org/docs/master/data.html#torch.utils.data.TensorDataset) to provide a known format to our dataset; in combination with [DataLoader](http://pytorch.org/docs/master/data.html#torch.utils.data.DataLoader), it will handle batching, shuffling, and other dataset iteration functions.

You can create data with TensorDataset by passing in feature and target tensors. Then create a DataLoader as usual.
```
data = TensorDataset(feature_tensors, target_tensors)
data_loader = torch.utils.data.DataLoader(data, 
                                          batch_size=batch_size)
```

### Batching
Implement the `batch_data` function to batch `words` data into chunks of size `batch_size` using the `TensorDataset` and `DataLoader` classes.

>You can batch words using the DataLoader, but it will be up to you to create `feature_tensors` and `target_tensors` of the correct size and content for a given `sequence_length`.

For example, say we have these as input:
```
words = [1, 2, 3, 4, 5, 6, 7]
sequence_length = 4
```

Your first `feature_tensor` should contain the values:
```
[1, 2, 3, 4]
```
And the corresponding `target_tensor` should just be the next "word"/tokenized word value:
```
5
```
This should continue with the second `feature_tensor`, `target_tensor` being:
```
[2, 3, 4, 5]  # features
6             # target
```

In [3]:
from torch.utils.data import TensorDataset, DataLoader
import numpy as np

def batch_data(words, sequence_length, batch_size):
    """
    Batch the neural network data using DataLoader
    :param words: The word ids of the TV scripts
    :param sequence_length: The sequence length of each batch
    :param batch_size: The size of each batch; the number of sequences in a batch
    :return: DataLoader with batched data
    """
    # TODO: Implement function
    words_length = len(words)
    features = []
    targets = []
    for i in range(words_length - sequence_length):
        features.append(words[i : i + sequence_length])
        targets.append(words[i + sequence_length])
    features = torch.LongTensor(np.array(features))
    targets = torch.LongTensor(np.array(targets))
    
    data = TensorDataset(features, targets)
    data_loader = torch.utils.data.DataLoader(data, batch_size=batch_size, shuffle=True)
    
    # return a dataloader
    return data_loader

# there is no test for this function, but you are encouraged to create
# print statements and tests of your own
words = [1, 2, 3, 4, 5, 6, 7]
sequence_length = 2
batch_size = 5
batch_data(words, sequence_length, batch_size)

### Test your dataloader 

You'll have to modify this code to test a batching function, but it should look fairly similar.

Below, we're generating some test text data and defining a dataloader using the function you defined, above. Then, we are getting some sample batch of inputs `sample_x` and targets `sample_y` from our dataloader.

Your code should return something like the following (likely in a different order, if you shuffled your data):

```
torch.Size([10, 5])
tensor([[ 28,  29,  30,  31,  32],
        [ 21,  22,  23,  24,  25],
        [ 17,  18,  19,  20,  21],
        [ 34,  35,  36,  37,  38],
        [ 11,  12,  13,  14,  15],
        [ 23,  24,  25,  26,  27],
        [  6,   7,   8,   9,  10],
        [ 38,  39,  40,  41,  42],
        [ 25,  26,  27,  28,  29],
        [  7,   8,   9,  10,  11]])

torch.Size([10])
tensor([ 33,  26,  22,  39,  16,  28,  11,  43,  30,  12])
```

### Sizes
Your sample_x should be of size `(batch_size, sequence_length)` or (10, 5) in this case and sample_y should just have one dimension: batch_size (10). 

### Values

You should also notice that the targets, sample_y, are the *next* value in the ordered test_text data. So, for an input sequence `[ 28,  29,  30,  31,  32]` that ends with the value `32`, the corresponding output should be `33`.

In [4]:
# test dataloader

test_text = range(4, 46)
t_loader = batch_data(test_text, sequence_length=5, batch_size=10)

data_iter = iter(t_loader)
sample_x, sample_y = data_iter.next()

print(sample_x.shape)
print(sample_x)
print()
print(sample_y.shape)
print(sample_y)

torch.Size([10, 5])
tensor([[ 33,  34,  35,  36,  37],
        [ 31,  32,  33,  34,  35],
        [ 19,  20,  21,  22,  23],
        [  5,   6,   7,   8,   9],
        [ 16,  17,  18,  19,  20],
        [ 35,  36,  37,  38,  39],
        [ 40,  41,  42,  43,  44],
        [ 30,  31,  32,  33,  34],
        [ 38,  39,  40,  41,  42],
        [ 13,  14,  15,  16,  17]])

torch.Size([10])
tensor([ 38,  36,  24,  10,  21,  40,  45,  35,  43,  18])


---
## Build the Neural Network
Implement an RNN using PyTorch's [Module class](http://pytorch.org/docs/master/nn.html#torch.nn.Module). You may choose to use a GRU or an LSTM. To complete the RNN, you'll have to implement the following functions for the class:
 - `__init__` - The initialize function. 
 - `init_hidden` - The initialization function for an LSTM/GRU hidden state
 - `forward` - Forward propagation function.
 
The initialize function should create the layers of the neural network and save them to the class. The forward propagation function will use these layers to run forward propagation and generate an output and a hidden state.

**The output of this model should be the *last* batch of word scores** after a complete sequence has been processed. That is, for each input sequence of words, we only want to output the word scores for a single, most likely, next word.

### Hints

1. Make sure to stack the outputs of the lstm to pass to your fully-connected layer, you can do this with `lstm_output = lstm_output.contiguous().view(-1, self.hidden_dim)`
2. You can get the last batch of word scores by shaping the output of the final, fully-connected layer like so:

```
# reshape into (batch_size, seq_length, output_size)
output = output.view(batch_size, -1, self.output_size)
# get last batch
out = output[:, -1]
```

In [32]:
import torch.nn as nn

class RNN(nn.Module):
    
    def __init__(self, vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5):
        """
        Initialize the PyTorch RNN Module
        :param vocab_size: The number of input dimensions of the neural network (the size of the vocabulary)
        :param output_size: The number of output dimensions of the neural network
        :param embedding_dim: The size of embeddings, should you choose to use them        
        :param hidden_dim: The size of the hidden layer outputs
        :param dropout: dropout to add in between LSTM/GRU layers
        """
        super(RNN, self).__init__()
        # TODO: Implement function
        
        # set class variables
        self.output_size = output_size
        self.n_layers = n_layers
        self.hidden_dim = hidden_dim
        
        # define model layers
        self.embedding = nn.Embedding(vocab_size, embedding_dim)
        self.lstm = nn.LSTM(embedding_dim, hidden_dim, n_layers, dropout=dropout, batch_first=True)
        
        self.dropout = nn.Dropout(p=dropout)
        self.fc = nn.Linear(hidden_dim, output_size)
        self.sig = nn.Sigmoid()
    
    
    def forward(self, nn_input, hidden):
        """
        Forward propagation of the neural network
        :param nn_input: The input to the neural network
        :param hidden: The hidden state        
        :return: Two Tensors, the output of the neural network and the latest hidden state
        """
        # TODO: Implement function
        batch_size = nn_input.size(0)
        
        if train_on_gpu:
            embeds = self.embedding(torch.LongTensor(nn_input.tolist()).cuda())
        else:
            embeds = self.embedding(torch.LongTensor(nn_input.tolist()))
        
        lstm_out, hidden = self.lstm(embeds, hidden)
        
        lstm_out = lstm_out.contiguous().view(-1, self.hidden_dim)
        
#         output = self.dropout(lstm_out)
        
        output = self.fc(lstm_out)
        output = output.view(batch_size, -1, self.output_size)
        
#         output = self.sig(output)
        output = output[:, -1]

        # return one batch of output word scores and the hidden state
        return output, hidden
    
    
    def init_hidden(self, batch_size):
        '''
        Initialize the hidden state of an LSTM/GRU
        :param batch_size: The batch_size of the hidden state
        :return: hidden state of dims (n_layers, batch_size, hidden_dim)
        '''
        # Implement function
        
        # initialize hidden state with zero weights, and move to GPU if available
        weight = next(self.parameters()).data
        
        if (train_on_gpu):
            hidden = (weight.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda(),
                     weight.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda())
        else:
            
            hidden = (weight.new(self.n_layers, batch_size, self.hidden_dim).zero_(),
                     weight.new(self.n_layers, batch_size, self.hidden_dim).zero_())
        return hidden

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_rnn(RNN, train_on_gpu)

Tests Passed


### Define forward and backpropagation

Use the RNN class you implemented to apply forward and back propagation. This function will be called, iteratively, in the training loop as follows:
```
loss = forward_back_prop(decoder, decoder_optimizer, criterion, inp, target)
```

And it should return the average loss over a batch and the hidden state returned by a call to `RNN(inp, hidden)`. Recall that you can get this loss by computing it, as usual, and calling `loss.item()`.

**If a GPU is available, you should move your data to that GPU device, here.**

In [33]:

def forward_back_prop(decoder, decoder_optimizer, criterion, inp, target, hidden):
    """
    Forward and backward propagation on the neural network
    :param decoder: The PyTorch Module that holds the neural network
    :param decoder_optimizer: The PyTorch optimizer for the neural network
    :param criterion: The PyTorch loss function
    :param inp: A batch of input to the neural network
    :param target: The target output for the batch of input
    :return: The loss and the latest hidden state Tensor
    """
    
    # TODO: Implement Function
    
    clipping_threshold = 5
    
    # move data to GPU, if available
    if (train_on_gpu):
        inp, target = inp.cuda(), target.cuda()
    
    # perform backpropagation and optimization
    hidden = tuple([each.data for each in hidden])
    
    decoder.zero_grad()
    
    output, hidden = decoder(inp, hidden)
    loss = criterion(output.squeeze(), target)
    loss.backward()
    
    nn.utils.clip_grad_norm_(decoder.parameters(), clipping_threshold)
    decoder_optimizer.step()

    # return the loss over a batch and the hidden state produced by our model
    return loss.item(), hidden

# Note that these tests aren't completely extensive.
# they are here to act as general checks on the expected outputs of your functions
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_forward_back_prop(RNN, forward_back_prop, train_on_gpu)

Tests Passed


## Neural Network Training

With the structure of the network complete and data ready to be fed in the neural network, it's time to train it.

### Train Loop

The training loop is implemented for you in the `train_decoder` function. This function will train the network over all the batches for the number of epochs given. The model progress will be shown every number of batches. This number is set with the `show_every_n_batches` parameter. You'll set this parameter along with other parameters in the next section.

In [34]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

def train_rnn(rnn, batch_size, optimizer, criterion, n_epochs, show_every_n_batches=100):
    batch_losses = []
    
    rnn.train()

    print("Training for %d epoch(s)..." % n_epochs)
    for epoch_i in range(1, n_epochs + 1):
        
        # initialize hidden state
        hidden = rnn.init_hidden(batch_size)
        
        for batch_i, (inputs, labels) in enumerate(train_loader, 1):
            
            # make sure you iterate over completely full batches, only
            n_batches = len(train_loader.dataset)//batch_size
            if(batch_i > n_batches):
                break
            
            # forward, back prop
            loss, hidden = forward_back_prop(rnn, optimizer, criterion, inputs, labels, hidden)          
            # record loss
            batch_losses.append(loss)

            # printing loss stats
            if batch_i % show_every_n_batches == 0:
                print('Epoch: {:>4}/{:<4}  Loss: {}\n'.format(
                    epoch_i, n_epochs, np.average(batch_losses)))
                batch_losses = []

    # returns a trained rnn
    return rnn

### Hyperparameters

Set and train the neural network with the following parameters:
- Set `sequence_length` to the length of a sequence.
- Set `batch_size` to the batch size.
- Set `num_epochs` to the number of epochs to train for.
- Set `learning_rate` to the learning rate for an Adam optimizer.
- Set `vocab_size` to the number of uniqe tokens in our vocabulary.
- Set `output_size` to the desired size of the output.
- Set `embedding_dim` to the embedding dimension; smaller than the vocab_size.
- Set `hidden_dim` to the hidden dimension of your RNN.
- Set `n_layers` to the number of layers/cells in your RNN.
- Set `show_every_n_batches` to the number of batches at which the neural network should print progress.

If the network isn't getting the desired results, tweak these parameters and/or the layers in the `RNN` class.

In [35]:
# Data params
# Sequence Length
sequence_length = 8  # of words in a sequence
# Batch Size
batch_size = 100

# data loader - do not change
train_loader = batch_data(int_text, sequence_length, batch_size)

In [36]:
# Training parameters
# Number of Epochs
num_epochs = 10
# Learning Rate
learning_rate = 0.001

# Model parameters
# Vocab size
vocab_size = len(vocab_to_int)
# Output size
output_size = len(vocab_to_int)

# Embedding Dimension
embedding_dim = 300
# Hidden Dimension
hidden_dim = 512
# Number of RNN Layers
n_layers = 2

# Show stats for every n number of batches
show_every_n_batches = 500

### Train
In the next cell, you'll train the neural network on the pre-processed data.  If you have a hard time getting a good loss, you may consider changing your hyperparameters. In general, you may get better results with larger hidden and n_layer dimensions, but larger models take a longer time to train. 
> **You should aim for a loss less than 3.5.** 

You should also experiment with different sequence lengths, which determine the size of the long range dependencies that a model can learn.

In [37]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# create model and move to gpu if available
rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
if train_on_gpu:
    rnn.cuda()

# defining loss and optimization functions for training
optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
criterion = nn.CrossEntropyLoss()

# training the model
trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

# saving the trained model
helper.save_model('./save/trained_rnn', trained_rnn)
print('Model Trained and Saved')

Training for 10 epoch(s)...
Epoch:    1/10    Loss: 5.292046100616455

Epoch:    1/10    Loss: 4.716120320796967

Epoch:    1/10    Loss: 4.531137160778045

Epoch:    1/10    Loss: 4.424124583244324

Epoch:    1/10    Loss: 4.353179054737091

Epoch:    1/10    Loss: 4.286316655158997

Epoch:    1/10    Loss: 4.23763388967514

Epoch:    1/10    Loss: 4.232313882350922

Epoch:    1/10    Loss: 4.156796869754792

Epoch:    1/10    Loss: 4.138613595485687

Epoch:    1/10    Loss: 4.130388983726501

Epoch:    1/10    Loss: 4.101504022598267

Epoch:    1/10    Loss: 4.075398155689239

Epoch:    1/10    Loss: 4.0727679080963135

Epoch:    1/10    Loss: 4.021873130321502

Epoch:    1/10    Loss: 4.051315469264984

Epoch:    1/10    Loss: 4.050359299182892

Epoch:    2/10    Loss: 3.9279943097556713

Epoch:    2/10    Loss: 3.8162444400787354

Epoch:    2/10    Loss: 3.8257139139175416

Epoch:    2/10    Loss: 3.7965649623870847

Epoch:    2/10    Loss: 3.8236549739837646

Epoch:    2/10    Los

/opt/conda/lib/python3.6/site-packages/torch/serialization.py:193: UserWarning: Couldn't retrieve source code for container of type RNN. It won't be checked for correctness upon loading.
  "type " + obj.__name__ + ". It won't be checked "


Model Trained and Saved
Training for 10 epoch(s)...
Epoch:    1/10    Loss: 5.306327684402466

Epoch:    1/10    Loss: 4.731471960544586

Epoch:    1/10    Loss: 4.52402490901947



KeyboardInterrupt: 

### Question: How did you decide on your model hyperparameters? 
For example, did you try different sequence_lengths and find that one size made the model converge faster? What about your hidden_dim and n_layers; how did you decide on those?

**Answer:** (Write answer, here)

1. tried to have 50 batch size, did not work (started on loss 10) so i increased the batch size to 100
2. tried to have 250 hidden dim, did not work (started on loss 9.2) so i increased the hidden dimension to 500.
3. tried to have learning rate of 0.01. did not work since the loss oscillates significantly from initial loss of 9.3. changed to 0.001 (this really improved the loss - starting at 9.23 to 9.18 on initial ). 

__THEN I REALIZED THAT WE DO NOT NEED SIGMOID IN THIS CASE SO I RETRAINED THE NETWORK WITHOUT SIGMOID__

1. to the new configuration with the following parameters:
    Training parameters
    num_epochs = 10
    learning_rate = 0.0001
    vocab_size = 46367
    output_size = 46367
    embedding_dim = 400
    hidden_dim = 500
    n_layers = 2
    error decreased from 10 to 6 on the initial results but loss does not decrease dramatically.
    
2. changed learning rate from 0.0001 to 0.001. from loss of 6 on the previous iteration, it improved to 5.5 on the initial result. Sequence length is 10 in this case. It performed well but since for me, sequence length is a big factor towards convergence, i tried several values.
3. changed the sequence length from 10 to 16. Decreased the inital loss from 5.5 to 5.4 but error decreases slowly.
4. I decreased the sequence length from 16 to 8. The training had an initial loss of 5.6 and it is not converging faster. 
5. So I tried sequence length of 12. Initial loss is 5.7 and loss is decreasing very well.
6. I tried sequence length of 11. Initial loss is 5.4 and is converging faster.

__ALL OF THE COMBINATIONS OF CHANGES TO HYPERPARAMETERS DID NOT WORK SINCE THE ERROR SATURATES AT 4. ERROR SOMEHOW DECREASES BUT WITH A COUPLE OF n_batches, IT OSCILLATES AGAIN. THEREFORE I TRIED DECREASING THE LEARNING RATE HOPING THAT I CAN PREVENT IT WITH THE FOLLOWING VALUES.__
    
    sequence_length = 8  # of words in a sequence
    batch_size = 100
    num_epochs = 10
    learning_rate = 0.0001
    vocab_size = len(vocab_to_int)
    output_size = len(vocab_to_int)
    embedding_dim = 250
    hidden_dim = 512
    n_layers = 2
    show_every_n_batches = 500

__NOW HERE IS THE UPDATE: DROPOUT OF 0.5 DOES NOT HELP. SINCE WE ARE DEALING WITH PAST WORD RELATIONSHIPS, TURNING OFF EACH NODE WITH THE PROBABILITY OF 50% EVERY FORWARD PASS IS TOO HIGH JUST TO PREVENT OVERFITTING. I REMOVED THAT AND MOVED TO HAVE THE SAME HYPERPARAMETERS AS ABOVE, ONLY WITH LEARNING RATE REVERTED FROM 0.0001 TO 0.001.__

    sequence_length = 8  # of words in a sequence
    batch_size = 100
    num_epochs = 10
    learning_rate = 0.001
    vocab_size = len(vocab_to_int)
    output_size = len(vocab_to_int)
    embedding_dim = 250
    hidden_dim = 512
    n_layers = 2
    show_every_n_batches = 500

__QUESTION: HOW DID I DECIDE ABOUT THESE HYPERPARAMETERS__

As you can see, it is more of a trial and error based on the values used from the exercises before this project and observing its behavior in at least 10 n_batches. Like in the case of my use of sigmoid where it is not needed since in the generate function below, softmax is used. Also in the dropout, where in the previous exercise it did help, in this case it did not. I spent approximately 10 hours of GPU time just to tweak on values. But in cases like hidden dimensions and number of layers, I chose to adhere to the standards and augment based on their standard multiples or increments.

---
# Checkpoint

After running the above training cell, your model will be saved by name, `trained_rnn`, and if you save your notebook progress, **you can pause here and come back to this code at another time**. You can resume your progress by running the next cell, which will load in our word:id dictionaries _and_ load in your saved model by name!

In [38]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch
import helper
import problem_unittests as tests

_, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()
trained_rnn = helper.load_model('./save/trained_rnn')

## Generate TV Script
With the network trained and saved, you'll use it to generate a new, "fake" Seinfeld TV script in this section.

### Generate Text
To generate the text, the network needs to start with a single word and repeat its predictions until it reaches a set length. You'll be using the `generate` function to do this. It takes a word id to start with, `prime_id`, and generates a set length of text, `predict_len`. Also note that it uses topk sampling to introduce some randomness in choosing the most likely next word, given an output set of word scores!

In [39]:
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import torch.nn.functional as F

def generate(rnn, prime_id, int_to_vocab, token_dict, pad_value, predict_len=100):
    """
    Generate text using the neural network
    :param decoder: The PyTorch Module that holds the trained neural network
    :param prime_id: The word id to start the first prediction
    :param int_to_vocab: Dict of word id keys to word values
    :param token_dict: Dict of puncuation tokens keys to puncuation values
    :param pad_value: The value used to pad a sequence
    :param predict_len: The length of text to generate
    :return: The generated text
    """
    rnn.eval()
    
    # create a sequence (batch_size=1) with the prime_id
    current_seq = np.full((1, sequence_length), pad_value)
    current_seq[-1][-1] = prime_id
    predicted = [int_to_vocab[prime_id]]
    
    for _ in range(predict_len):
        if train_on_gpu:
            current_seq = torch.LongTensor(current_seq).cuda()
        else:
            current_seq = torch.LongTensor(current_seq)
        
        # initialize the hidden state
        hidden = rnn.init_hidden(current_seq.size(0))
        
        # get the output of the rnn
        output, _ = rnn(current_seq, hidden)
        
        # get the next word probabilities
        p = F.softmax(output, dim=1).data
        if(train_on_gpu):
            p = p.cpu() # move to cpu
         
        # use top_k sampling to get the index of the next word
        top_k = 5
        p, top_i = p.topk(top_k)
        top_i = top_i.numpy().squeeze()
        
        # select the likely next word index with some element of randomness
        p = p.numpy().squeeze()
        word_i = np.random.choice(top_i, p=p/p.sum())
        
        # retrieve that word from the dictionary
        word = int_to_vocab[word_i]
        predicted.append(word)     
        
        # the generated word becomes the next "current sequence" and the cycle can continue
        current_seq = np.roll(current_seq, -1, 1)
        current_seq[-1][-1] = word_i
    
    gen_sentences = ' '.join(predicted)
    
    # Replace punctuation tokens
    for key, token in token_dict.items():
        ending = ' ' if key in ['\n', '(', '"'] else ''
        gen_sentences = gen_sentences.replace(' ' + token.lower(), key)
    gen_sentences = gen_sentences.replace('\n ', '\n')
    gen_sentences = gen_sentences.replace('( ', '(')
    
    # return all the sentences
    return gen_sentences

### Generate a New Script
It's time to generate the text. Set `gen_length` to the length of TV script you want to generate and set `prime_word` to one of the following to start the prediction:
- "jerry"
- "elaine"
- "george"
- "kramer"

You can set the prime word to _any word_ in our dictionary, but it's best to start with a name for generating a TV script. (You can also start with any other names you find in the original text file!)

In [42]:
# run the cell multiple times to get different results!
gen_length = 400 # modify the length to your preference
prime_word = 'kramer' # name for starting the script

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
pad_word = helper.SPECIAL_WORDS['PADDING']
generated_script = generate(trained_rnn, vocab_to_int[prime_word + ':'], int_to_vocab, token_dict, vocab_to_int[pad_word], gen_length)
print(generated_script)

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:46: UserWarning: RNN module weights are not part of single contiguous chunk of memory. This means they need to be compacted at every call, possibly greatly increasing memory usage. To compact weights again call flatten_parameters().


kramer: riverside park.

jerry: oh, well, i guess i could get some sleep.

jerry: oh, you know, i think i'm going to go to the movies next week?

jerry: well, i guess i'll see you later.(she hangs up)

george: what, you didn't have to go to that bachelor auction, and now you want to be a communist, i was just standing there.

jerry: well, i was thinking i didn't realize it, so i don't know what i'm doing in my life, i know, i'm going to be honest with him anymore.

elaine: well, i guess i should.

jerry: well, i don't know how you feel about it.

jerry: what do you mean" assman"?

susan: yeah.

george:(to himself) i didn't know you wanted to know what it was.

elaine: well i don't have to be a little nervous.

jerry: what?

newman: well, i'm not really looking at a drug central.

kramer: hey, hey, hey hey! hey!

jerry: hey.

kramer: hey, hey buddy, i got it.(snaps fingers)

kramer: yeah, but, you know, the usual.

jerry: well, i guess i should be a lesbian convertible?

george: yeah.



#### Save your favorite scripts

Once you have a script that you like (or find interesting), save it to a text file!

In [43]:
# save script to a text file
f =  open("generated_script_1.txt","w")
f.write(generated_script)
f.close()

# The TV Script is Not Perfect
It's ok if the TV script doesn't make perfect sense. It should look like alternating lines of dialogue, here is one such example of a few generated lines.

### Example generated script

>jerry: what about me?
>
>jerry: i don't have to wait.
>
>kramer:(to the sales table)
>
>elaine:(to jerry) hey, look at this, i'm a good doctor.
>
>newman:(to elaine) you think i have no idea of this...
>
>elaine: oh, you better take the phone, and he was a little nervous.
>
>kramer:(to the phone) hey, hey, jerry, i don't want to be a little bit.(to kramer and jerry) you can't.
>
>jerry: oh, yeah. i don't even know, i know.
>
>jerry:(to the phone) oh, i know.
>
>kramer:(laughing) you know...(to jerry) you don't know.

You can see that there are multiple characters that say (somewhat) complete sentences, but it doesn't have to be perfect! It takes quite a while to get good results, and often, you'll have to use a smaller vocabulary (and discard uncommon words), or get more data.  The Seinfeld dataset is about 3.4 MB, which is big enough for our purposes; for script generation you'll want more than 1 MB of text, generally. 

# Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_tv_script_generation.ipynb" and save another copy as an HTML file by clicking "File" -> "Download as.."->"html". Include the "helper.py" and "problem_unittests.py" files in your submission. Once you download these files, compress them into one zip file for submission.